### ISSUE TITLE
Upload and download module should not show HTML files

### ISSUE THREAD

**Issue by jcameron** [webmin/webmin#1907]
Seems potentially risky if the root user shows an un-trusted HTML file in the browser..

However, I did some tests and this would only happen if the "Show in browser if possible?" option is set to Yes. @iliajie does that match your experience?

![image](https://github.com/webmin/webmin/assets/102706/a09a75a7-4cc3-4047-b54f-2130ca0d7396)

**Comment by iliajie** [comment]
Also, it's worth mentioning that detection of file type can be tricky. For example `&guess_mime_type()` sub and `file` command produce slightly different results on the file with mixed content.

For example, if we create `text.txt` file and place in it:

```
hello world
```

It will be recognized as `text/plain` in both cases. However, if we add to that file:

```
<script>alert(1234)</script>
```

Then `&guess_mime_type()` would still call that file `text/plain`, while `file` command output would return `HTML document, ASCII text`.

Although, luckily (and correctly) a browser would also treat that kind of file as plain text.

**Comment by jcameron** [comment]
Surely all that matters is what MIME type we set in the response header?

**Comment by iliajie** [comment]
> Surely all that matters is what MIME type we set in the response header?

Oh, actually, right! There is a better solution, I think we can always force `text/plain` header for any kind of text files. What do you think of this patch?

https://github.com/webmin/webmin/commit/5e9f137adfc28226121a351b2f78cc8db73630a1

**Comment by jcameron** [comment]
No I don't like that approach, the regexp is too broad as it would match all kinds of MIME types that include the word "text". Please undo it, and I'll work on a safer solution, perhaps using the `filter_javascript` function.

**Comment by jcameron** [comment]
I think this is a better fix : https://github.com/webmin/webmin/commit/ab44aa802343f2ae4b2a95852436e12a48603b0c

**Comment by jcameron** [comment]
Oh god, I didn't know that SVG files could contain javascript!!

That patch won't entirely work, because filtering has to be done on the entire file at once in case some JS crosses the buffer boundary. But otherwise it's good ...

**Comment by jcameron** [comment]
Yeah, `read_file_contents` would work just as well.

I'm not sure what other types are potentially dangerous to display..

**Comment by chris001** [comment]
[Types from microsoft](https://learn.microsoft.com/en-us/deployedge/microsoft-edge-security-downloads-interruptions) and [chromium](https://source.chromium.org/chromium/chromium/src/+/main:components/safe_browsing/core/resources/download_file_types.asciipb;drc=af17ad3f07c1d8a24381eb7669bec0c2ffb86521).

**Comment by iliajie** [comment]
> Yeah, `read_file_contents` would work just as well.

Alright, although I think there is a better solution. Have a look at [this patch](https://github.com/webmin/webmin/commit/d5a3eebe85a00a66247178eb2a5a428a040e7102).



> I'm not sure what other types are potentially dangerous to display.
>> [Types from microsoft](https://learn.microsoft.com/en-us/deployedge/microsoft-edge-security-downloads-interruptions) and [chromium](https://source.chromium.org/chromium/chromium/src/+/main:components/safe_browsing/core/resources/download_file_types.asciipb;drc=af17ad3f07c1d8a24381eb7669bec0c2ffb86521).

I will check.

**Comment by iliajie** [comment]
Jamie, alright. I have changed existing regex to be a bit more catchy. We generally should be fine. However, I am still not 100% sure if we're covering it all. For example, what do we do with these kind of PDF files, e.g.:

[pdf-alert-1.pdf](https://github.com/webmin/webmin/files/11496056/pdf-alert-1.pdf)

**Comment by iliajie** [comment]
We could technically test for:

```
<<
/Type /Action
/S /JavaScript
```

.. and force download instead?

**Comment by iliajie** [comment]
> Maybe the only safe option is to always download, and remove the option to show in the browser entirely?

Probably not, because there are a lot of use cases when an admin deals with trusted files.

We can filter out in a similar way to JavaScript all PDFs for:

```
<<
/Type /Action
/S /JavaScript
...
>>
```

**Comment by jcameron** [comment]
Editing files via regexps is a risky thing to do safely though ... there's many ways it can go wrong and either break a file, or miss some malicious JS that avoids our regexp.

**Comment by iliajie** [comment]
I know but we don't have many options.

**Comment by jcameron** [comment]
Well we could just not show those files in the browser at all ... just limit to files that are 100% safe or can be safely filtered, like GIF, JPG and text.

**Comment by iliajie** [comment]
Is JPG totally save though?

**Comment by jcameron** [comment]
Yes (as long as there's not some JPG parsing bug in the browser).

Unsure about PNG ....

**Comment by iliajie** [comment]
> Well we could just not show those files in the browser at all ... just limit to files that are 100% safe or can be safely filtered, like GIF, JPG and text.

Once we already filter JavaScript it's no brainer to filter it our from PDF as well. Check this patch - https://github.com/webmin/webmin/commit/7b37cedaf6c303a5304ee61ae4c147a0281cdeff.

**Comment by jcameron** [comment]
Are you sure it's safe to use the same function to filter JS from PDF and HTML?

**Comment by iliajie** [comment]
> Are you sure it's safe to use the same function to filter JS from PDF and HTML?

Well, even thought it'd be safe most likely, after thinking more over it, I didn't find it to be great code-wise.

If we had a separate _filter-lib_, then I would most definitely created a separate sub for it. As in our case, so not to add an extra rarely used function to the global scope, I modified existing `filter_javascript` sub.

Also, I decided not to risk by filtering and showing altered PDF or XML content. This change https://github.com/webmin/webmin/commit/3de93296a93c3843c7f6bc9f31c51996bd1de57b will take care of the following:

  1. If PDF or SVG, for example after trying the filter was altered, then we force downloading it, otherwise if no dangerous content was found, just show it normally
  2. If HTML was not filtered and there was no JS found, then just display it as is, however, if there was JS found, pre-insert the banner saying so, as silently changing and displaying the HTML file isn't a user-friendly thing, so e.g.:
    <img width="565" alt="image" src="https://github.com/webmin/webmin/assets/4426533/5a3527f9-ddba-49da-b344-bfc449272960">

I think this is the best solution to maintain the balance between usability and security.

What ya think?

**Comment by jcameron** [comment]
Can you do this change as a PR so we can discuss it before committing? I think this solution is just too complex and potentially risky, and I really don't like the idea of inserting a message into the HTML! 

Better to just do the simplest and safest thing, which is to only show file types in the browser if we are 100% sure they are safe (text, GIF, PNG and JPG). Actually, I suspect that most users of the file manager and upload/download modules just want to fetch a file to their PC, rather than viewing it in the browser.

**Comment by swelljoe** [comment]
I was about to close this, seeing the merged PR about it, but I don't think that PR is safe. Nothing that shows potentially untrusted HTML (or anything that can contain JavaScript) in a Webmin session can possibly be safe. That's asking for XSS, and much worse. We shouldn't offer to display HTML from the filesystem in a Webmin session. I was going to say it could go to a frame without the Webmin session, but I'm not sure if that's possible? But, displaying random HTML in a Webmin session is so dangerous I can't believe regexes would even be suggested as a solution.

**Comment by iliaross** [comment]
Thanks, Joe! Agreed! Fixed here https://github.com/webmin/webmin/commit/11fafb738d07ecdecb60de665d4062f3f9a56639.
